lang/src/org/partiql/lang/eval/physical/operators/Accumulator.kt

```diff
@@ -0,0 +1,184 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package org.partiql.lang.eval.physical.operators
+
+import org.partiql.lang.domains.PartiqlPhysical
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.eval.ExprAggregator
+import org.partiql.lang.eval.ExprValue
+import org.partiql.lang.eval.ExprValueFactory
+import org.partiql.lang.eval.NaturalExprValueComparators
+import org.partiql.lang.eval.createUniqueExprValueFilter
+import org.partiql.lang.eval.errNoContext
+import org.partiql.lang.eval.isUnknown
+import org.partiql.lang.eval.numberValue
+import org.partiql.lang.util.bigDecimalOf
+import org.partiql.lang.util.div
+import org.partiql.lang.util.plus
+import java.math.BigDecimal
+
+internal sealed class Accumulator(
+    internal open val filter: (ExprValue) -> Boolean
+) : ExprAggregator {
+    companion object {
+        internal fun create(funcName: String, quantifier: PartiqlPhysical.SetQuantifier, valueFactory: ExprValueFactory): Accumulator {
+            val filter = when (quantifier) {
+                is PartiqlPhysical.SetQuantifier.Distinct -> createUniqueExprValueFilter()
+                is PartiqlPhysical.SetQuantifier.All -> { _: ExprValue -> true }
+            }
+            return when (funcName.trim().toLowerCase()) {
+                "min" -> AccumulatorMin(valueFactory, filter)
+                "max" -> AccumulatorMax(valueFactory, filter)
+                "avg" -> AccumulatorAvg(valueFactory, filter)
+                "count" -> AccumulatorCount(valueFactory, filter)
+                "sum" -> AccumulatorSum(valueFactory, filter)
+                "group_as" -> AccumulatorGroupAs(valueFactory, filter)
+                else -> throw IllegalArgumentException("Unsupported aggregation function: $funcName")
+            }
+        }
+    }
+
+    override fun next(value: ExprValue) {
+        if (value.isUnknown() || filter.invoke(value).not()) return
+        nextValue(value)
+    }
+
+    abstract fun nextValue(value: ExprValue)
+}
+
+internal class AccumulatorSum(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var sum: Number? = null
+
+    override fun nextValue(value: ExprValue) {
+        checkIsNumberType(funcName = "SUM", value = value)
+        if (sum == null) sum = 0.0
+        this.sum = value.numberValue() + this.sum!!
+    }
+
+    override fun compute(): ExprValue {
+        val totalSum = sum?.let { bigDecimalOf(it) }
+        return totalSum?.exprValue(valueFactory) ?: valueFactory.nullValue
+    }
```

Comment:
Nit: It seems like the sum function will always return `Decimal` Type. In the evaluating compiler, sum may return an integer or float type.

Reply:
Thanks! Looking into this!

Reply:
You're totally right -- submitting a revision once tests show passing.